lib/alex_koin/slack_commands.ex

```diff
@@ -53,88 +53,112 @@ defmodule AlexKoin.SlackCommands do
 
   defp update_user_info(user, _user_info), do: user
 
-  def get_balance(wallet) do
-    wallet.balance
-  end
-
-  def get_coins(_wallet) do
+  defp get_user_wallet(%User{id: user_id}) do
+    Wallet
+    |> Repo.get_by(user_id: user_id)
+    |> case do
+      nil -> {:error, :not_found}
+      wallet -> {:ok, wallet}
+    end
   end
 
-  def create_coin(user, created_by_user, reason) do
-    user_wallet = Wallet |> Repo.get_by(user_id: user.id)
-
-    new_coin = %Coin{
-      origin: reason,
-      mined_by_id: user.id,
-      hash: UUID.uuid1(),
-      created_by_user_id: created_by_user.id,
-      wallet_id: user_wallet.id
-    }
-
-    {:ok, coin} = Repo.insert(new_coin)
-
-    # Now we create the initial transaction to set the coin up
-    {:ok, _txn} = transfer_coin(coin, user_wallet, user_wallet, "Initial creation.")
-
-    coin
+  def create_coin(user = %User{}, created_by_user = %User{}, reason) do
+    Multi.new()
+    |> Multi.run(:wallet, fn _repo, _changes ->
+      get_user_wallet(user)
+    end)
+    |> Multi.run(:new_coin, fn _repo, %{wallet: wallet} ->
+      Coins.create_coin(%{
+        origin: reason,
+        mined_by_id: user.id,
+        hash: UUID.uuid1(),
+        created_by_user_id: created_by_user.id,
+        wallet_id: wallet.id
+      })
+    end)
+    |> Multi.run(:transfer_coin, fn _repo, %{new_coin: new_coin, wallet: wallet} ->
+      # Now we create the initial transaction to set the coin up
+      transfer_coin(new_coin, wallet, wallet, "Initial creation.")
+    end)
+    |> Repo.transaction()
+    |> case do
+      {:ok, %{new_coin: new_coin}} -> {:ok, new_coin}
+      err -> err
+    end
   end
 
-  def remove_coins(from_wallet, amount) do
-    coins = Repo.all(Coin.for_wallet(from_wallet, amount))
-    Enum.each(coins, fn c -> Coins.delete_coin(c) end)
-
-    AlexKoin.Account.update_wallet(from_wallet, %{balance: from_wallet.balance - amount})
+  def remove_coins(from_wallet = %Wallet{balance: balance}, amount) do
+    Multi.new()
+    |> Multi.run(:delete_coins, fn _, _ ->
+      coins =
+        from_wallet
+        |> Coin.for_wallet(amount)
+        |> Repo.all()
+
+      if length(coins) == amount do
+        Enum.each(coins, &Coins.delete_coin(&1))
```

Comment:
Does this work? I think it doesn't since you're trying to create a bunch of queries without each being a Multi.run. I ran into this issue with the transfer logic, and couldn't wrap my head around how to do that properly.

Reply:
Hmm, yeah I'll need to restructure a bit. You can pass around the multi so that might be the way forward here.

Reply:
I think my new way is better here.